gateway-api:  Supports Addresses provided by the Gateway API specification

chaunceyjiang opened the pull request:
Fixes: #32865

Cilium Gateway supports Addresses provided by the Gateway API specification https://gateway-api.sigs.k8s.io/api-types/gateway/?h=addresses



```release-note
Cilium Gateway supports Addresses provided by the Gateway API specification
```

sayboras commented on .github/workflows/conformance-gateway-api.yaml:
```diff
@@ -131,7 +131,7 @@ jobs:
         run: |
           echo sha=${{ steps.default_vars.outputs.sha }} >> $GITHUB_OUTPUT
           
-          EXEMPT_FEATURES="GatewayStaticAddresses,HTTPRouteParentRefPort,HTTPRouteResponseHeaderModification"
```
Sadly, this change will not be reflected in CI on forked repo, so you might need to run the test locally and share the result here, or push your  change to cilium/cilium branch,  and  create another branch.

Add my previous comment here for reference 

https://github.com/cilium/cilium/pull/33042#issuecomment-2160814887

chaunceyjiang committed:
gateway-api: Supports Addresses provided by the Gateway API specification

Fixes: #32865

Cilium Gateway supports Addresses provided by the Gateway API specification https://gateway-api.sigs.k8s.io/api-types/gateway/?h=addresses

Signed-off-by: chaunceyjiang <chaunceyjiang@gmail.com>

chaunceyjiang commented on operator/pkg/gateway-api/conformance_test.go:
```diff
@@ -4,13 +4,23 @@
 )
 
+var (
+	usableNetworkAddressesEnv   = "GATEWAY_API_CONFORMANCE_USABLE_NETWORK_ADDRESSES"
```
My plan is to pass these two variables to `CONFORMANCE TEST` through environment variables.

youngnick replied:
Using environment variables for this makes sense to me.

youngnick commented on operator/pkg/gateway-api/conformance_test.go:
```diff
@@ -32,6 +42,38 @@ import (
+		skipTests = append(skipTests, string(features.SupportGatewayStaticAddresses))
+	} else {
+		var addressType = v1.IPAddressType
+		usableNetworkAddressesStr := os.Getenv(usableNetworkAddressesEnv)
```
Nit: This call to `os.Getenv()` could be before the `if`, and then there would be only one call rather than two. Same for `unusableNetworkAddressesEnv`.

Not a super big deal here, but if this call was expensive or in a hot path, it would matter more.

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -172,13 +186,16 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+		if len(gw.Spec.Addresses) != 0 {
+			// This annotation must be set through gateway.spec.addresses, otherwise it will be deleted.
+			delete(svc.Annotations, annotation.LBIPAMIPKeyAlias)
+		}
```
This is actually a breaking change, because if someone was setting this annotation on a created Service resource manually, it will now be removed.

I'd suggest we log a warning about deprecated label use and say that this will change in the future to removing this label if no `spec.Addresses` are set.

Something like:
```
DEPRECATED: The service <servnicenamespace/<servicename> is setting an IP address using the annotation <annotation>. These should be set using the spec.addresses field in Gateway objects instead. At a future date this annotation will be removed from Gateway-created Services if no spec.addresses are set.
```

That's a long message though, suggestions welcomed.

Then we need a followup issue to come back and change this to delete the annotation instead.

rouke-broersma replied:
What if I use the infrastructure stanza for this, will that also not be supported anymore?

youngnick replied:
If we make this remove those annotations, then yes, there would be a problem. Practically, I think that if we do make the breaking change here, we will need to check the labels and annotations and do something (either put a warning in the status, or reject outright) Gateways that use infrastructure to set this label.

Really good point, thanks for that.

youngnick commented on operator/pkg/gateway-api/gateway_status.go:
```diff
@@ -11,24 +11,24 @@ import (
 // setGatewayProgrammed inserts or updates the Programmed condition for the provided Gateway resource.
-func setGatewayProgrammed(gw *gatewayv1.Gateway, ready bool, msg string) *gatewayv1.Gateway {
-	gw.Status.Conditions = merge(gw.Status.Conditions, gatewayStatusProgrammedCondition(gw, ready, msg))
+func setGatewayProgrammed(gw *gatewayv1.Gateway, ready bool, msg string, reason gatewayv1.GatewayConditionReason) *gatewayv1.Gateway {
```
Great work adding Reason here as well. This is a good change.

youngnick reviewed (changes requested):
Mostly LGTM, with one small change to make this change non-breaking for now.

sayboras commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -172,13 +190,12 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 }
 
-func (r *gatewayReconciler) ensureService(ctx context.Context, desired *corev1.Service) error {
+func (r *gatewayReconciler) ensureService(ctx context.Context, desired *corev1.Service, gw *gatewayv1.Gateway) error {
```
nit: this new gw param is not used

chaunceyjiang replied:
Done.

sayboras reviewed (approved):
This feature is highly asked by community, thanks a lot for your contribution.

I have one nit as per below, also other related docs (if required) can be done in separate PR as well.

youngnick reviewed (approved):
LGTM, nice work @chaunceyjiang. Let's get this in and documented.
